visualize_GT4D.py

```diff
@@ -0,0 +1,404 @@
+    amp_audio_path = base_dir / content / "audio" / "micamp" / f"micamp_{filename}.wav"
+    midi_path = base_dir / content / "midi" / f"midi_{filename}.mid"
+
+    # Verify required video files exist.
+    if not ego_video_path.exists():
+        raise FileNotFoundError(f"Ego video file not found: {ego_video_path}")
+    if not exo_video_path.exists():
+        raise FileNotFoundError(f"Exo video file not found: {exo_video_path}")
+
+    # Extract audio from video files.
+    ego_audio, egosr = get_audio_from_video(ego_video_path)
+    exo_audio, exosr = get_audio_from_video(exo_video_path)
+
+    if egosr != exosr:
+        raise ValueError("Sample rates for ego and exo videos do not match.")
+
+    # Load additional audio files.
+    try:
+        di_audio, _ = librosa.load(str(di_audio_path), sr=egosr)
+        amp_audio, _ = librosa.load(str(amp_audio_path), sr=egosr)
+    except Exception as e:
+        logging.error(f"Error loading additional audio files: {e}")
+        raise
+
+    # Load MIDI data and construct tablature matrix.
+    try:
+        midi_data = pretty_midi.PrettyMIDI(str(midi_path))
+        for instrument in midi_data.instruments:
+            if instrument.name not in Config.GUITAR_STRINGS:
+                raise ValueError(f"Invalid instrument name: {instrument.name}")
+    except Exception as e:
+        logging.error(f"Error loading MIDI file {midi_path}: {e}")
+        raise
+
+    tab_matrix = build_tab_matrix(midi_data, time_resolution=Config.TIME_RESOLUTION)
+
+    # Prepare dictionary of signals and log pairwise lags.
+    signals = {
+        "ego": ego_audio,
+        "exo": exo_audio,
+        "direct_input": di_audio,
+        "mic_amp": amp_audio,
+    }
+
+    # Load and resize video clips.
+    try:
+        ego_clip = VideoFileClip(str(ego_video_path))
+        exo_clip = VideoFileClip(str(exo_video_path))
+    except Exception as e:
+        logging.error(f"Error loading video clips: {e}")
+        raise
+
+    # Resize the clips to have the same height.
+    target_height = min(ego_clip.h, exo_clip.h)
+    ego_clip_resized = ego_clip.resize(height=target_height)
+    exo_clip_resized = exo_clip.resize(height=target_height)
+    
+    # Arrange the clips side by side.
+    combined_clip = clips_array([[ego_clip_resized, exo_clip_resized]])
+    
+    full_duration = combined_clip.duration
+
+    # Validate and compute the time range.
+    if not (0 <= start_pct < end_pct <= 100):
+        raise ValueError("start_pct must be >= 0, end_pct <= 100, and start_pct < end_pct.")
+    start_time = full_duration * (start_pct / 100)
+    end_time = full_duration * (end_pct / 100)
+    segment_duration = end_time - start_time
+
+    # Dynamically generate the output filename if not provided
+    if output_filename is None:
+        output_filename = f"output_{filename}_start{int(start_pct)/100:.2f}_end{int(end_pct)/100:.2f}.mp4"
```

Comment:
the `output_filename` should include the `directory` basename to avoid overwriting videos of the same data but from different participants in the dataset.